Apromore-Extras/APMLogModule/src/main/java/org/apromore/apmlog/filter/PLog.java

```diff
@@ -778,27 +787,22 @@ private PTrace getPTraceByImmutableIndex(int index) {
     public List<PTrace> getCustomPTraceList() {
         BitSet currentBS = this.validTraceIndexBS;
 
-        List<PTrace> theCusPTraceList = new ArrayList<>();
-
-        List<ATrace> immutableTraces = apmLog.getTraceList();
+        Map<String, List<PTrace>> idPTraces = pTraceList.stream()
+                .parallel()
+                .collect(Collectors.groupingBy(PTrace::getCaseId));
 
-        int mutableIndex = 0;
+        List<PTrace> theCusPTraceList = originalPTraceList;
 
-        for (ATrace aTrace : immutableTraces) {
-            PTrace pt = null;
-            if (!currentBS.get(aTrace.getImmutableIndex())) {
-                pt = new PTrace(mutableIndex, aTrace, this);
-                pt.setValidEventIndexBS(new BitSet(aTrace.getEventSize()));
+        for (PTrace pTrace : theCusPTraceList) {
+            if (currentBS.get(pTrace.getImmutableIndex())) {
+                pTrace.setValidEventIndexBS(idPTraces.get(pTrace.getCaseId()).get(0).getValidEventIndexBitSet());
```

Comment:
Can you guarantee that .setValidEventIndexBS is never empty, or that the IndexOutOfBounds exception will be handled when .get(0) is called?